### ISSUE TITLE
Saves can't be loaded due to ImportError (cPickle)

### ISSUE THREAD

**Issue by Victorius-Astral** [renpy/renpy#3495]
Saves made on older Renpy versions can't be loaded in the latest stable. Due to an ImportError of the 'revertable' module, that doesn't exist. When I added the placeholder .py module, it said that the module has no attribute 'RevertableList'. It should be Renpy's modified list object made for rollback. Could add a compatibility patch for this?

EDIT:
I've just checked the nightly branch, and it was readded there, can you add this to a stable, though? I don't exactly know the current state of nightly, but the last time I tried (two/three months ago) I had some issues (Android if I remember correctly) so I went back to the stable.

```
I'm sorry, but an uncaught exception occurred.

While running game code:
  File "renpy/common/00action_file.rpy", line 462, in __call__
    renpy.load(fn)
ImportError: No module named revertable

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "renpy/common/_layout/screen_main_menu.rpym", line 28, in script
    python hide:
  File "renpy/ast.py", line 928, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 2245, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "renpy/common/_layout/screen_main_menu.rpym", line 28, in <module>
    python hide:
  File "renpy/common/_layout/screen_main_menu.rpym", line 35, in _execute_python_hide
    ui.interact()
  File "renpy/ui.py", line 298, in interact
    rv = renpy.game.interface.interact(roll_forward=roll_forward, **kwargs)
  File "renpy/display/core.py", line 3325, in interact
    repeat, rv = self.interact_core(preloads=preloads, trans_pause=trans_pause, pause=pause, pause_start=pause_start, **kwargs)
  File "renpy/display/core.py", line 4160, in interact_core
    rv = root_widget.event(ev, x, y, 0)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/transition.py", line 48, in event
    return self.new_widget.event(ev, x, y, st) # E1101
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/screen.py", line 727, in event
    rv = self.child.event(ev, x, y, st)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1323, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 273, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1323, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 273, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 273, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/behavior.py", line 983, in event
    return handle_click(self.clicked)
  File "renpy/display/behavior.py", line 918, in handle_click
    rv = run(action)
  File "renpy/display/behavior.py", line 330, in run
    return action(*args, **kwargs)
  File "renpy/common/00action_file.rpy", line 462, in __call__
    renpy.load(fn)
  File "renpy/loadsave.py", line 770, in load
    roots, log = loads(location.load(filename))
  File "renpy/loadsave.py", line 63, in loads
    return cPickle.loads(s)
ImportError: No module named revertable

Windows-10-10.0.22000
Ren'Py 7.4.11.2266
Astral Lust 0.2.2c
Thu May 12 09:00:54 2022
```

**Comment by Victorius-Astral** [comment]
```
I'm sorry, but an uncaught exception occurred.

While running game code:
  File "renpy/common/00action_file.rpy", line 462, in __call__
    renpy.load(fn)
AttributeError: 'module' object has no attribute 'RevertableList'

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "renpy/common/_layout/screen_main_menu.rpym", line 28, in script
    python hide:
  File "renpy/ast.py", line 928, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 2245, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "renpy/common/_layout/screen_main_menu.rpym", line 28, in <module>
    python hide:
  File "renpy/common/_layout/screen_main_menu.rpym", line 35, in _execute_python_hide
    ui.interact()
  File "renpy/ui.py", line 298, in interact
    rv = renpy.game.interface.interact(roll_forward=roll_forward, **kwargs)
  File "renpy/display/core.py", line 3325, in interact
    repeat, rv = self.interact_core(preloads=preloads, trans_pause=trans_pause, pause=pause, pause_start=pause_start, **kwargs)
  File "renpy/display/core.py", line 4160, in interact_core
    rv = root_widget.event(ev, x, y, 0)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/transition.py", line 48, in event
    return self.new_widget.event(ev, x, y, st) # E1101
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/screen.py", line 727, in event
    rv = self.child.event(ev, x, y, st)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1323, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 273, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1323, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 273, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1102, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 273, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/behavior.py", line 983, in event
    return handle_click(self.clicked)
  File "renpy/display/behavior.py", line 918, in handle_click
    rv = run(action)
  File "renpy/display/behavior.py", line 330, in run
    return action(*args, **kwargs)
  File "renpy/common/00action_file.rpy", line 462, in __call__
    renpy.load(fn)
  File "renpy/loadsave.py", line 770, in load
    roots, log = loads(location.load(filename))
  File "renpy/loadsave.py", line 63, in loads
    return cPickle.loads(s)
AttributeError: 'module' object has no attribute 'RevertableList'

Windows-10-10.0.22000
Ren'Py 7.4.11.2266
Astral Lust 0.2.2c
Thu May 12 09:06:05 2022
```

**Comment by Gouvernathor** [comment]
What placeholder .py module are you talking about ?
In what version of renpy was the game saved initially, and where is it attempted to be loaded ? A save from an older release should always be loadable in a newer release or nightly, but not the other way around : saves from nightlies can fail (and that's why you're not supposed to distribute games from them) and saves from newer releases can fail in older ones.

**Comment by Victorius-Astral** [comment]
Placeholder is what all placeholders are. A file having no real value. It was an empty module with just Hello World. 

To put the issue here straight up to your face, Release 7.4.11.2266 from 2th December 2021 does not have revertable.py in  Ren'Py/renpy folder. I demoted from 7.5.0.22051201n where the file was present to the stable again, and the file revertable.py having RevertableList class was not there. It's either corruption of the updater package of an installer. It can also be that the file was never present there. It might be present in a .zip downloadable from the site, idk. But I can confirm that updating to the newest stable twice deleted the needed revertable.py from Renpy/renpy both times, unless my launcher is having issues.

I don't remember which version I was using a year ago, but it's irrelevant, it's updating to stable that deletes revertable.py, which was present in previous versions, and is present on nightly.

I provide a screenshot of the renpy folder on nightly and stable, I also noticed rollback.py was deleted too, perhaps all .py starting with 'r' were deleted. I didn't notice rollback issues, though. Hope it helps.

![1](https://user-images.githubusercontent.com/78815540/168066018-35b2099b-cdd4-4093-8a49-65fd1a4759c4.jpg)
![2](https://user-images.githubusercontent.com/78815540/168066022-250423c2-01fc-4dbf-b627-96a44d6b83fc.jpg)
.

**Comment by Gouvernathor** [comment]
What you're describing is normal : revertable.py was not a part of renpy 7.4-, it was added during 7.5. It should not be there in 7.4, and it's good that it's not there. In renpy 7.4-, RevertableList is in renpy.python iirc.

My hypothesis as to why the save/load bug you described triggered, is that you saved your game in 7.5, and tried to load it in a previous version of renpy, which as I said is not and cannot be supported.

**Comment by Victorius-Astral** [comment]
This might be the case, I got it from a player asking for support, and when I tried to load his save I got this. Since it was not there before 7.5, then it's the only option then. I have had 7.5 used to compile one release, it must be from there, I just didn't expect that it was a new module since lists working with rollback were present since long ago. At least I never noticed then not working with it, instead OrderedDict from collections module didn't work with rollback, even if it was defined with 'default'

Thanks for the help.
